hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/observer/ObservableImpl.java

```diff
@@ -0,0 +1,297 @@
+import com.hazelcast.ringbuffer.StaleSequenceException;
+import com.hazelcast.ringbuffer.impl.RingbufferProxy;
+import com.hazelcast.spi.exception.DistributedObjectDestroyedException;
+import com.hazelcast.spi.impl.executionservice.ExecutionService;
+
+import javax.annotation.Nonnull;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+
+import static com.hazelcast.jet.impl.JobRepository.INTERNAL_JET_OBJECTS_PREFIX;
+
+public class ObservableImpl<T> implements Observable<T> {
+
+    /**
+     * Prefix of all topic names used to back {@link Observable} implementations,
+     * necessary so that such topics can't clash with regular topics used
+     * for other purposes.
+     */
+    public static final String JET_OBSERVABLE_NAME_PREFIX = INTERNAL_JET_OBJECTS_PREFIX + "observables.";
+
+    /**
+     * Constant ID to be used as a {@link ProcessorMetaSupplier#getTags()
+     * PMS tag key} for specifying when a PMS owns an {@link Observable} (ie.
+     * is the entity populating the {@link Observable} with data).
+     */
+    public static final String OWNED_OBSERVABLE = "owned_observable";
+
+    private final ConcurrentMap<UUID, RingbufferListener<T>> listeners = new ConcurrentHashMap<>();
+    private final String name;
+    private final HazelcastInstance hzInstance;
+    private final Consumer<Observable<T>> onDestroy;
+    private final ILogger logger;
+
+    public ObservableImpl(String name, HazelcastInstance hzInstance, Consumer<Observable<T>> onDestroy, ILogger logger) {
+        this.name = name;
+        this.hzInstance = hzInstance;
+        this.onDestroy = onDestroy;
+        this.logger = logger;
+    }
+
+    @Nonnull @Override
+    public String name() {
+        return name;
+    }
+
+    @Nonnull @Override
+    public UUID addObserver(@Nonnull Observer<T> observer) {
+        UUID id = UuidUtil.newUnsecureUUID();
+        RingbufferListener<T> listener = new RingbufferListener<>(name, id, observer, hzInstance, logger);
+        listeners.put(id, listener);
+        listener.next();
+        return id;
+    }
+
+    @Override
+    public void removeObserver(@Nonnull UUID registrationId) {
+        RingbufferListener<T> listener = listeners.remove(registrationId);
+        if (listener == null) {
+            throw new IllegalArgumentException(
+                    String.format("No registered observer with registration ID %s", registrationId));
+        } else {
+            listener.cancel();
+        }
+    }
+
+    @Override
+    public void destroy() {
+        listeners.keySet().forEach(this::removeObserver);
+        listeners.clear();
+
+        // destroy the underlying ringbuffer
+        hzInstance.getRingbuffer(ringbufferName(name)).destroy();
+        onDestroy.accept(this);
+        //TODO (PR-1729): we shouldn't allow usages of this class, after it has been destroyed
```

Comment:
Apparently I'm able to forget about TODO's many times. Putting a note in the PR, maybe that helps...